kafka-0.10.x/src/main/scala/monix/kafka/KafkaProducerSink.scala

```diff
@@ -39,29 +40,26 @@ final class KafkaProducerSink[K, V] private (
 
   require(parallelism >= 1, "parallelism >= 1")
 
-  def createSubscriber(cb: Callback[Throwable, Unit], s: Scheduler) = {
+  def createSubscriber(
+    cb: Callback[Throwable, Unit],
+    s: Scheduler): (Subscriber[Seq[ProducerRecord[K, V]]], AssignableCancelable.Multi) = {
     val out = new Subscriber[Seq[ProducerRecord[K, V]]] { self =>
-      implicit val scheduler = s
+      implicit val scheduler: Scheduler = s
       private[this] val p = producer.memoize
       private[this] var isActive = true
 
-      private def sendAll(batch: Seq[ProducerRecord[K, V]]): Seq[Task[Option[RecordMetadata]]] =
-        for (record <- batch) yield {
-          try p.value.send(record)
-          catch { case NonFatal(ex) => Task.raiseError(ex) }
-        }
-
       def onNext(list: Seq[ProducerRecord[K, V]]): Future[Ack] =
         self.synchronized {
           if (!isActive) Stop
           else {
             val sendTask: Task[Seq[Option[RecordMetadata]]] =
               if (parallelism == 1)
-                Task.sequence(sendAll(list))
+                Task.traverse(list)(p.value().send(_))
               else {
-                val batches = list.sliding(parallelism, parallelism)
-                val tasks = for (b <- batches) yield Task.gather(sendAll(b))
-                Task.sequence(tasks.toList).map(_.flatten)
+                for {
+                  s <- Semaphore[Task](parallelism)
+                  res <- Task.wander(list)(r => s.withPermit(p.value().send(r)))
```

Comment:
The problem with this implementation of `wanderN` is that it will start all of the tasks at the beginning which can consume a lot of memory if we have very big number. We will get proper `wanderN` in Monix soon: https://github.com/monix/monix/pull/954 but I'm not sure what's more efficient without measuring. Could be nice to double check if `list.sliding` + sending in parallel isn't more efficient when dealing with sth like 30k msg/s

But it's fine if you don't have time for it, we can leave it as is

Reply:
> Could be nice to double check if list.sliding + sending in parallel isn't more efficient when dealing with sth like 30k msg/s

I think I can check this with same benchmarks

Reply:
Added some more benchmarks to this repo:
https://github.com/voidconductor/monix-kafka-benchmark

As I can see - semaphore version is faster, but I don't know how do I create 30k msg/s load on producer

Reply:
I've never used Scalameter but I guess we could just increase the load and it would come naturally :D

Or ensure it with `Observable` doing something like `bufferTimedWithPressure`

Reply:
Thanks, @Avasil 
I did more measurements, and indeed, when huge chunks (30k+) are sent in parallel - wander with semaphore creates 100% cpu load and write speed really unstable (5-100 MB/s), while sliding iterator consistently sends at ~55 MB/s with around 5-10% cpu load

Probably, when proper implementation is merged to monix it should be used here instead of semaphore

Reply:
Yeah, we can exchange it quite easily, I will do it when updating for `3.0.0`

Many thanks for your investigation!